fabtests/component/sock_test.c

```diff
@@ -0,0 +1,253 @@
+	ret = getaddrinfo(server_addr, NULL, NULL, &res);
+	if (ret) {
+		printf("getaddrinfo failed\n");
+		return ret;
+	}
+
+	memcpy(server_sock, res->ai_addr, res->ai_addrlen);
+	server_addr_len = res->ai_addrlen;
+
+	switch (server_sock->ss_family) {
+	case AF_INET:
+		in = (struct sockaddr_in *) server_sock;
+		in->sin_port = PORT;
+		break;
+	case AF_INET6:
+		in6 = (struct sockaddr_in6 *) server_sock;
+		in6->sin6_port = PORT;
+		break;
+	default:
+		printf("Unsupported Address family\n");
+		ret = -1;
+		break;
+	}
+
+	*sock = socket(server_sock->ss_family, SOCK_STREAM, 0);
+	if (sock < 0)
+		return -1;
+
+	ret = setsockopt(*sock, SOL_SOCKET, SO_REUSEADDR, (char *) &optval, sizeof(optval));
+	if (ret) {
+		printf("error setting socket options\n");
+		return ret;
+	}
+
+	ret = bind(*sock, (struct sockaddr *)server_sock, server_addr_len);
+
+	return ret;
+}
+
+int server_create(int connections, struct sockaddr_storage server_sock, int sock, int** clients)
+{
+	int new_sock, ret, i;
+
+	ret = listen(sock, connections);
+	if (ret) {
+		printf ("listen failed\n");
+		return ret;
+	}
+
+	*clients = calloc(connections, sizeof(int));
+	if (!*clients)
+		return -1;
+
+	for (i = 0; i < connections; i++) {
+		new_sock = accept(sock, NULL, NULL);
+		if (new_sock < 0) {
+			printf("error during server init\n");
+			return -1;
+		}
+
+		(*clients)[i] = new_sock;
+		
+	}
+	return 0;
+}
+
+int client_connect(int connections, struct sockaddr_storage server_addr, int sock, int **clients)
+{
+	int i, ret, new_sock;
+	int optval = 1;
+
+	*clients = calloc(connections, sizeof(int));
+
+
+	for (i = 0; i < connections; i++) {
+
+		new_sock = socket(server_addr.ss_family, SOCK_STREAM, 0);
+		if (new_sock < 0)
+			return -1;
+
+		ret = setsockopt(new_sock, SOL_SOCKET, SO_REUSEADDR, (char *) &optval, sizeof(optval));
+		if (ret) {
+			printf("error setting socket options\n");
+			return ret;
+		}
+
+		ret = connect(new_sock, (struct sockaddr *)&server_addr, server_addr_len);
+		if (ret)
+			return ret;
+
+		(*clients)[i] = new_sock;
+	}
+
+	return ret;
+}
+
+double time_select(int connections, int server_sock, int *clients) 
+{
+	fd_set select_set;
+	int i, max_sock;
+	struct timeval timeout;
+	struct timespec start;
+	struct timespec stop;
+
+	timeout.tv_sec = 0;
+	timeout.tv_usec = 0;
+
+	FD_ZERO(&select_set);
+	FD_SET(server_sock, &select_set);
+
+	max_sock = server_sock;
+
+	for (i = 0; i < connections; i++) {
+		FD_SET(clients[i], &select_set);
+
+		if (clients[i] > max_sock)
+			max_sock = clients[i];
+
+	}
+
+	clock_gettime(CLOCK_MONOTONIC, &start);
+	select(max_sock + 1, &select_set, NULL, NULL, &timeout);
+	clock_gettime(CLOCK_MONOTONIC, &stop);
+
+	return (stop.tv_nsec + stop.tv_sec * 1000000000) - (start.tv_nsec + start.tv_sec * 1000000000);
+}
+
+double time_poll(int connections, int server_sock, int *clients)
+{
+	struct pollfd *poll_set;
+	int i;
+	struct timespec start;
+	struct timespec stop;
+
+	poll_set = malloc(sizeof(struct pollfd) * connections);

```

Comment:
check for failures

Reply:
and please delete poll_set before the function returns otherwise it would leak memory.